Express/src/services/collectionService.ts

```diff
@@ -0,0 +1,168 @@
+import { BookWithAuthorsAndGenres } from "../../types/BookAuthorGenre";
+import { BookWithAuthorsAndGenres } from "../../types/BookAuthorGenre";
+import * as dataAcess from "../data-access/collectionDataAccess";
+
+export async function fetchPublicCollections() {
+  const publicCollections = dataAcess.getPublicCollections();
+  if (!publicCollections.length) {
+    throw { status: 404, message: "No public collections found" };
+  }
+  return publicCollections;
+}
+
+export async function fetchUserCollections(userId: string) {
+  const userCollections = await dataAcess.getUserCollections(userId);
+  if (!userCollections.length) {
+    throw { status: 404, message: "No collection found" };
+  }
+  return userCollections;
+}
+
+export async function fetchCollectionDetails(
+  collectionId: number,
+  userId?: string
+) {
+  const collection = await dataAccess.getCollectionByID(collectionId);
+  if (collection == null) {
+    throw { status: 404, message: "Collection not found" };
+  }
+
+  const isOwner = collection.userId == userId;
+  if (!collection.public && isOwner === false) {
+    throw { status: 403, message: "Forbidden: Access denied" };
+  }
+
+  const booksInCollection = await dataAccess.getBooksInCollection(collectionId);
+
+  const booksWithDetails: BookWithAuthorsAndGenres[] = [];
+  for (const book of booksInCollection) {
+    const [authors, genres] = Promise.all([
+      dataAccess.getBookAuthors(book.id),
+      dataAccess.getBookGenres(book.id),
+    ]);
+
+    booksWithDetails.push({
+      ...book,
+      authors: authors.map((author) => author.names),
+      genres: genres.map((genre) => genre.genre),
+    });
+  }
+
+  return {
+    title: collection.name,
+    books: booksWithDetails,
+    isOwner,
+    public: collection.public,
+  };
+}
+
+export async function createNewCollection(
+  name: string,
+  userId: number,
+  publicStatus: boolean
+) {
+  const existingCollection = await dataAccess.checkExistingCollections(
+    userId,
+    name
+  );
+  if (existingCollection == undefined) {
+    throw {
+      status: 409,
+      message: "User already has a collection with this name",
+    };
+  }
+
+  const newCollection = await dataAccess.createCollection(
+    name,
+    userId,
+    publicStatus
+  );
+  return newCollection;
+}
+
+export async function addBookToUserCollection(
+  collectionId: number,
+  bookId: number,
+  userId: string
+) {
+  const collection = await dataAccess.getCollectionById(collectionId);
+  if (!collection) {
+    throw { status: 404, message: "Collection not found" };
+  }
+  if (collection.user_id !== userId) {
+    throw { status: 403, message: "Forbidden: Access denied" };
+  }
+  const bookExists = await dataAccess.checkBookExist(bookId);
+  if (!bookExists) {
+    throw { status: 404, message: "Book not found" };
+  }
+  const bookInCollection = dataAccess.checkBookInCollection(
```

Comment:
`dataAccess.checkBookInCollection` grąžina `Promise` objektą, todėl jam reikalingas `await`
```suggestion
  const bookInCollection = await dataAccess.checkBookInCollection(
```